### PR TITLE
`Development`: Use object mapper and streams to improve and simplify the code

### PR DESCRIPTION
**Pull request by krusche**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

We aim to replace the Google [gson](https://github.com/google/gson) library with the Jackson ObjectMapper for parsing JSON and XML. This change will help us maintain consistency and simplify the server code. Additionally, we are gradually transitioning from using classes to records for our DTOs.

### Description

This PR reduces the reliance on gson and replaces it with the ObjectMapper. We focus on the structure oracle generator classes used for generating structural tests in programming exercises. Additionally we replace classes related to Legal Documents, which were used as DTOs, with records.

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Student
- 1 Programming Exercise with structural tests

Best tested locally, or on a test server where you are allowed to change the Imprint / Privacy Statement as admin
1. Log in to Artemis as admin
2. Navigate to Server Administration
3. Under Imprint set a new text and save, same for Privacy Statement
4. Click on Imprint / Privacy in the footer to see if the new text appears
5. Create a programming exercise with structural tests (there should be some per default)
6. Check that they pass for solution and fail for template
7. Participate in the programing exercise as a student
8. See that the tests are passing as expected

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [imprint.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Client-Tests/app/core/legal/imprint.component.ts.html) | 88.23% | ✅ |
| [privacy.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Client-Tests/app/core/legal/privacy.component.ts.html) | 95% | ✅ |
| [legal-document.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/legal-document.model.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [LegalDocumentService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/LegalDocumentService.html) | 100% | ✅ |
| [DataExportExerciseCreationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.export/DataExportExerciseCreationService.html) | 93% | ✅ |
| [JavaClassDiff.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.util.structureoraclegenerator/JavaClassDiff.html) | 87% | ✅|
| [JavaClassDiffSerializer.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.util.structureoraclegenerator/JavaClassDiffSerializer.html) | 81% | ✅ |
| [OracleGenerator.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.util.structureoraclegenerator/OracleGenerator.html) | 83% | ✅ |
| [SerializerUtil.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.util.structureoraclegenerator/SerializerUtil.html) | 69% | ✅ |
| [AdminImprintResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminImprintResource.html) | 100% | ✅ |
| [AdminPrivacyStatementResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminPrivacyStatementResource.html) | 100% | ✅ |
| [ImprintDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/ImprintDTO.html) | 100% | ✅ |
| [PrivacyStatementDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/PrivacyStatementDTO.html) | 100% | ✅ |
| [PublicImprintResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.open/PublicImprintResource.html) | 100% | ✅ |
| [PublicPrivacyStatementResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6357/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.open/PublicPrivacyStatementResource.html) | 100% | ✅ |

### CODE REVIEW

**Review by pzdr7 — DISMISSED** [review]
Tested on TS3.

- Changed and viewed the imprint; worked for both English and German ✅ 
- Created a programming exercise with the default template
  - Template: 0% ✅ 
  - Solution: 100% ✅ 
  - Participation: Tests pass and fail as expected ✅

File: src/test/java/de/tum/in/www1/artemis/PrivacyStatementResourceIntegrationTest.java [hunk 1]
```diff
@@ -19,27 +21,26 @@
 import org.mockito.MockedStatic;
 import org.springframework.http.HttpStatus;
 import org.springframework.security.test.context.support.WithMockUser;
+import org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper;
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/test/java/de/tum/in/www1/artemis/PrivacyStatementResourceIntegrationTest.java]: done

**Review by JohannesWt — APPROVED** [review]
Code LGTM 👍

File: src/test/java/de/tum/in/www1/artemis/exercise/programming/ProgrammingSubmissionAndResultGitlabJenkinsIntegrationTest.java [hunk 2]
```diff
@@ -11,6 +11,8 @@
 import java.util.Map;
 import java.util.stream.Stream;
 
+import jakarta.validation.constraints.NotNull;
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/test/java/de/tum/in/www1/artemis/exercise/programming/ProgrammingSubmissionAndResultGitlabJenkinsIntegrationTest.java]: done

File: src/test/java/de/tum/in/www1/artemis/architecture/ArchitectureTest.java [hunk 3]
```diff
@@ -293,9 +297,7 @@ void testNoRestControllersImported() {
         @Override
         public void check(JavaClass item, ConditionEvents events) {
             item.getDirectDependenciesFromSelf().stream().map(Dependency::getTargetClass).filter(targetClass -> targetClass.isAnnotatedWith(RestController.class))
-                    .forEach(targetClass -> {
-                        events.add(violated(item, "%s imports the RestController %s".formatted(item.getName(), targetClass.getName())));
-                    });
+                    .forEach(targetClass -> events.add(violated(item, "%s imports the RestController %s".formatted(item.getName(), targetClass.getName()))));
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/test/java/de/tum/in/www1/artemis/architecture/ArchitectureTest.java:300]: I'm not sure what you mean here.

File: src/main/java/de/tum/in/www1/artemis/service/export/DataExportExerciseCreationService.java [hunk 4]
```diff
@@ -338,7 +338,7 @@ private void createResultsAndComplaintFiles(Submission submission, Path outputDi
                         if (feedback.getTestCase() != null && feedback.getTestCase().getTestName() != null) {
                             resultScoreAndFeedbacks.append(feedback.getTestCase().getTestName()).append("\t");
                         }
-                        // getText() contains, e.g., the file and line number for programming exercises
+                        // text() contains, e.g., the file and line number for programming exercises
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/export/DataExportExerciseCreationService.java:341]: unrelated
```suggestion
                        // getText() contains, e.g., the file and line number for programming exercises
```

File: src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/JavaClassDiff.java [hunk 5]
```diff
@@ -333,7 +334,7 @@ else if (solutionList == null || templateList == null) {
         if (solutionList.size() != templateList.size()) {
             return false;
         }
-        return solutionList.containsAll(templateList);
+        return new HashSet<>(solutionList).containsAll(templateList);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/JavaClassDiff.java:337]: Why is this necessary?

Comment (laadvo) [reply src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/JavaClassDiff.java:337]: IntelliJ marks `solutionList.containsAll(templateList)` yellow saying "Call to 'list.containsAll(collection)' may have poor performance" and suggests using `new HashSet<>(solutionList).containsAll(templateList);` instead.

**Review by marlon-luca-bu — APPROVED** [review]
Code LGTM

**Review by JohannesWt — DISMISSED** [review]
Code reapprove (no changes)

File: src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/JavaClassDiffSerializer.java [hunk 6]
```diff
@@ -36,77 +39,78 @@ class JavaClassDiffSerializer {
...
-                continue;
-            }
-            JsonObject attributeJSON = SerializerUtil.createJsonObject(attribute.getName(), new HashSet<>(attribute.getModifiers()), attribute, attribute.getAnnotations());
-            attributeJSON.addProperty("type", attribute.getType().getValue());
-            attributesJSON.add(attributeJSON);
-        }
+    ArrayNode serializeAttributes() {
+        ArrayNode attributesJSON = mapper.createArrayNode();
+
+        // Using Stream API for a cleaner approach
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/JavaClassDiffSerializer.java:91]:
```suggestion
```
This comment does not add anything useful

File: src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/JavaClassDiffSerializer.java [hunk 7]
```diff
@@ -36,77 +39,78 @@ class JavaClassDiffSerializer {
...
-        JsonArray enumsJSON = new JsonArray();
+    ArrayNode serializeEnums() {
+        ArrayNode enumsJSON = mapper.createArrayNode();
 
-        for (JavaField javaEnum : javaClassDiff.enumsDiff) {
-            if (!isElementToIgnore(javaEnum)) {
-                enumsJSON.add(javaEnum.getName());
-            }
-        }
+        // Stream API for filtering and processing
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/JavaClassDiffSerializer.java:112]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/JavaClassDiffSerializer.java [hunk 8]
```diff
@@ -117,52 +121,50 @@ JsonArray serializeEnums() {
...
+
+        javaClassDiff.methodsDiff.stream().filter(method -> !isElementToIgnore(method)).forEach(method -> methodsJSON.add(createMethodJson(method)));
+
         return methodsJSON;
     }
 
+    private ObjectNode createMethodJson(JavaMethod method) {
+        ObjectNode methodJSON = SerializerUtil.createJsonObject(method.getName(), new HashSet<>(method.getModifiers()), method, method.getAnnotations());
+
+        // Directly add parameters and return type without checking for emptiness
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/JavaClassDiffSerializer.java:161]:
```suggestion
        // No need to check for isEmpty before adding; let the serializer utility methods decide how to handle empty collections
```

File: src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java [hunk 9]
```diff
@@ -69,50 +75,40 @@ public class OracleGenerator {
...
-            }
-
-            // If we are dealing with interfaces, the types diff already has all the information we need
-            // So we do not need to do anything more
-            JavaClassDiffSerializer serializer = new JavaClassDiffSerializer(javaClassDiff);
-            diffJSON.add("class", serializer.serializeClassProperties());
-            if (!javaClassDiff.methodsDiff.isEmpty()) {
-                diffJSON.add("methods", serializer.serializeMethods());
-            }
+        solutionToTemplateMapping.entrySet().stream().map(entry -> generateDiffJSON(entry.getKey(), entry.getValue())).filter(java.util.Optional::isPresent)
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java:81]:
```suggestion
        solutionToTemplateMapping.entrySet().stream().map(entry -> generateDiffJSON(entry.getKey(), entry.getValue())).filter(Optional::isPresent)
```

File: src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java [hunk 10]
```diff
@@ -69,50 +75,40 @@ public class OracleGenerator {
...
+                .map(java.util.Optional::get).forEach(structureOracleJSON::add);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java:82]:
```suggestion
                .map(Optional::get).forEach(structureOracleJSON::add);
```

File: src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java [hunk 11]
```diff
@@ -69,50 +75,40 @@ public class OracleGenerator {
...
+        if (!javaClassDiff.enumsDiff.isEmpty()) {
+            diffJSON.set("enumValues", serializer.serializeEnums());
+        }
+        if (!javaClassDiff.constructorsDiff.isEmpty()) {
+            diffJSON.set("constructors", serializer.serializeConstructors());
         }
 
-        return prettyPrint(structureOracleJSON);
+        log.debug("Generated JSON for '{}'.", solutionType.getCanonicalName());
+        return java.util.Optional.of(diffJSON);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java:111]:
```suggestion
        return Optional.of(diffJSON);
```

File: src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java [hunk 12]
```diff
@@ -140,22 +143,22 @@ private static Map<JavaClass, JavaClass> generateSolutionToTemplateMapping(Path
...
-            solutionToTemplateMapping.put(solutionClass, null);
-
-            for (JavaClass templateClass : templateClasses) {
-                if (solutionClass.getSimpleName().equals(templateClass.getSimpleName()) && templateClass.getPackageName().equals(solutionClass.getPackageName())) {
-                    // If a template class with the same name and package gets found, then replace the empty template with the real one.
-                    solutionToTemplateMapping.put(solutionClass, templateClass);
-                    break;
-                }
-            }
+            String qualifiedName = solutionClass.getPackageName() + "." + solutionClass.getSimpleName();
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java:159]:
```suggestion
            String qualifiedName = solutionClass.getName();
```
```suggestion
            String qualifiedName = solutionClass. getCanonicalName();
```
Why doesn't one of those work?

File: src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java [hunk 13]
```diff
@@ -140,22 +143,22 @@ private static Map<JavaClass, JavaClass> generateSolutionToTemplateMapping(Path
...
+            JavaClass templateClass = templateClassMap.getOrDefault(qualifiedName, null);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java:160]:
```suggestion
            JavaClass templateClass = get(qualifiedName);
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by JohannesStoehr — APPROVED** [review]

**Review by Jan-Thurner — APPROVED** [review]
Tested on TS3. Works as expected.

**Review by DominikRemo — APPROVED** [review]
Code LGTM

**Review by pzdr7 — APPROVED** [review]
Tested again on TS3. Imprint, privacy statement, and programming exercises still work as expected.
